confgenerator/confgenerator.go

```diff
@@ -377,6 +408,37 @@ func generateOtelExporters(exporters map[string]*config.MetricsExporter, pipelin
 	return stackdriverList, exportNameMap, nil
 }
 
+func generateOtelProcessors(processors map[string]*config.MetricsProcessor, pipelines map[string]*config.MetricsPipeline) ([]*otel.ExcludeMetrics, map[string]string, error) {
+	excludeMetricsList := []*otel.ExcludeMetrics{}
+	processorNameMap := make(map[string]string)
+	excludemetricsProcessorFactories, err := extractOtelProcessorFactories(processors)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, pID := range config.SortedKeys(pipelines) {
+		p := pipelines[pID]
+		for _, processorID := range p.ProcessorIDs {
+			if _, ok := processorNameMap[processorID]; ok {
+				continue
+			}
+			if p, ok := excludemetricsProcessorFactories[processorID]; ok {
+				var metricPrefixes []string
```

Comment:
This kind of string manipulation should be done in the template. Just put the raw metric prefix slice into the config struct.

(This is especially important when we have workload metrics to drop, and it should make things shorter to boot.)

Reply:
Can you clarify this a bit more?

Reply:
Are you implying me to do something similar to this: 
![image](https://user-images.githubusercontent.com/8354694/123483783-49edbe80-d5d5-11eb-8004-597e1853e093.png)

also, do we want to do this now. this is not the only place we massage the data in confgenerator. Removing the massage data into struct can be done in a separate PR.

Reply:
Recreating this comment thread after rebase.

> also, do we want to do this now. this is not the only place we massage the data in confgenerator. Removing the massage data into struct can be done in a separate PR.

Just because existing code is bad is not a good reason to write more bad code. :)

I think there's three different string manipulations that need to happen between the OA config and the OT config:

1. `agent.googleapis.com/` needs to be stripped as a prefix (with a giant TODO to fix this when workload metrics are supported; actually you should probably make sure that it validates that this starts with `agent.googleapis.com/` until workload metrics are supported)
2. Globs need to be turned into regexes
3. Lists of strings needs to be formatted as valid YAML

The third one should definitely happen in the template. Something like

```
metric_names:
{{range .MetricNames}}
  - {{printf "%q" .}}
{{end}}
```

That properly handles the escaping you need, which your current code doesn't even attempt.

The other two are a bit more ambiguous about where they should be; it might be reasonable to keep them here for now but 1 eventually goes away with workload metrics and 2 would be best as a filter function in the template. Your `strings.TrimPrefix` call is fine to remove `agent.googleapis.com`, but you need to actually handle the glob and properly escape the special characters in the string. That would be something like:

```
var literals []string
for _, p := range strings.Split(glob, "*") {
  literals = append(literals, regexp.QuoteMeta(p))
}
metricNames = append(metricNames, fmt.Sprintf(`^%s$`, strings.Join(literals, `.*`)))
```

to replace lines 429-430 in this file.

Escaping is really important, parsing globs is important, and generating YAML in one place instead of multiple is important.

Reply:
`for _, p := range strings.Split(glob, "*")` 
This line seems predicting the prefixes could be something like `agent.googleapis.com\c*u\*` which i think a good idea implementation-wise. As we might support more generic field other than `metric_prefixes` to something like `metric_globs` then the same code can be used for both cases?
Thus fixes applied

Reply:
I suggested that because Joe specifically Decided that these needed to be globs and not just prefixes, yeah.

Reply:
it's now metrics_pattern